File: Orbis_Watchmen_Retail_API/js/orbisapi_watchmen_auth.js
import * as dotenv from "dotenv";
import { v4 as uuidv4 } from "uuid";
import jose from "node-jose";
import axios from "axios";
import fs from "fs";

const conf = dotenv.config().parsed;
const { JWE, JWK } = jose;
const keystore = JWK.createKeyStore();
const keyfile = `${conf.ACCESSKEY}.pem`;

const payload = JSON.stringify({
  iss: "watchman API",
  iat: Math.floor(Date.now() / 1000),
  exp: Math.floor((Date.now() + 60 * 1000) / 1000),
  nbf: Math.floor(Date.now() / 1000),
  jti: uuidv4(),
});

(async () => {
  // Read public key file
  const public_key = fs.readFileSync(keyfile);
  const key = await keystore.add(public_key, "pem", { kid: conf.ACCESSKEY });

  try {
    //// Create JWE token
    var jwetoken = await JWE.createEncrypt(
      {
        format: "compact",
        fields: {
          alg: "RSA-OAEP-256",
          kid: conf.ACCESSKEY,
          enc: "A128GCM",
        },
      },
      key
    )
      .update(payload)
      .final();

    //// Get token
    const res = await axios({
      method: "post",
      url: `${conf.HOSTNAME}/v1/oms/token`,
      data: {
        jwe: jwetoken,
      },
      headers: {
        "Content-type": "application/json",
      },
    });
    console.log(res.data);

    //// Query
    // const querydata = await axios.get(conf.HOSTNAME + "/v1/users/search/accounts?criteria=YIKANG&lookup=true", {
    //   headers: {
    //     Authorization: "Bearer " + res.data.content,
    //   },
    // });
    // console.log(querydata.data);
  } catch (err) {
    console.log(err);
  }
})();

File: Orbis_Watchmen_Retail_API/python/orbisapi_watchmen_auth.py
import requests
import os
import json
import base64
from jwcrypto import jwk, jwe
from uuid import uuid4
from datetime import datetime
from dotenv import dotenv_values

if __name__ == '__main__':
    config = dotenv_values(".env")

    with open(f'{config["ACCESSKEY"]}.pem', 'rb') as f:
        public_key = jwk.JWK.from_pem(f.read())

    protected_header = {
        "alg": "RSA-OAEP-256",
        "enc": "A128GCM",
        "kid": config["ACCESSKEY"],
    }

    payload = {
        'iss': 'watchman API',
        'iat': int(datetime.now().timestamp()),
        'exp': int(datetime.now().timestamp()) + 60,
        'nbf': int(datetime.now().timestamp()),
        'jti': str(uuid4()),
    }

    # Create jwe token
    jwetoken = jwe.JWE(
        json.dumps(payload), recipient=public_key, protected=protected_header
    )
    enc = jwetoken.serialize(compact=True)

    ## Get token
    r = requests.post(
        f'{config["HOSTNAME"]}/v1/oms/token',
        headers={
            'Content-type': 'application/json',
        },
        data=json.dumps({'jwe': enc}),
    )
    token = json.loads(r.text)['content']
    print(json.loads(r.text))

    ## Request
    # r = requests.get(
    #     f'{config["HOSTNAME"]}/v1/users/search/accounts?criteria=YIKANG&lookup=true',
    #     headers={
    #         'Content-type': 'application/json',
    #         'Authorization': f'Bearer {token}',
    #     },
    # )
    # print(json.loads(r.text))
